### DISCUSSION TITLE
Possibility of using Extended Property Pattern in With Expression?

### DISCUSSION THREAD

**Discussion by KanaHayama** [dotnet/csharplang#6897] [Language Ideas] [answered]
Hi, I am a person trying to use record classes to manage program states recently.
I am trapped in an annoying situation when I am using them.

I define many record classes and combine them together.
I nest some record classes in some other record classes to make the data structure clear.
An I found, to modify a property of nested record class, I have to write lines of codes.

Here is a contrived example for the problem.
First, we have some record classes.
```CSharp
#nullable enable

record CharacterState{
    public int Health { get; init; }
}

record GameState{
    public CharacterState Player { get; init; } = new CharacterState();
}
```

When I am holding a reference to an outer record class, I want to mutate a property of a nested record.
I first write something like this.
```CSharp
static GameState Update(GameState state) {
    // Stage: Recover health
    var result = state with {
        Player = state.Player with {
            Health = state.Player.Health + 1,
        },
    };
    return result;
}
```
The code of mutating a single property will become longer if I have more layers of nested record classes.

I remember there is a language feature that allows me to write code like this.
```CSharp
if (state is { Player.Health: 0 }) {
    //do something
}
```
It is called [Extended Property Patterns](https://learn.microsoft.com/en-us/dotnet/csharp/language-reference/proposals/csharp-10.0/extended-property-patterns) added in C#10.
Essentially, the nested properties can be concatenated by dots.

I feel my use case is similar.
All classes are record classes so they should support [With Expression](https://learn.microsoft.com/en-us/dotnet/csharp/language-reference/operators/with-expression).
So, I tried this.
```CSharp
static GameState Update(GameState state) {
    // Stage: Recover health
    var result = state with {
        Player.Health = state.Player.Health + 1,
    };
    return result;
}
```
I got a compiler error.

My code has a lot of places that mutate one property of nested record class, so simple operations become exceptionally long, eventually it becomes a noise when I am programming.
I tried to search for solutions on the Internet to see if my programming paradigm or usage is wrong, but I found no useful information.
I also did not find an github issue talking about this, so I want to ask if there is any solution for it.

Thank you for providing helps.

**Comment by HaloFour** [comment, marked as answer]
See: #4587

**Reply by KanaHayama** [reply]
Thank you for pointing me to the existing discussions.
